vespalib/src/vespa/vespalib/stllike/hashtable.h

```diff
@@ -141,39 +141,38 @@ class hashtable : public hashtable_base
         typedef Value* pointer;
         typedef std::forward_iterator_tag iterator_category;
 
-        iterator(hashtable * hash, next_t start) : _hash(start), _subNode(start), _hashTable(hash) {
-            advanceToNextValidHash();
-        }
-        iterator(hashtable * hash, next_t start, next_t subNode) : _hash(start), _subNode(subNode), _hashTable(hash) { }
-        Value & operator * ()  const { return _hashTable->get(_subNode); }
-        Value * operator -> () const { return & _hashTable->get(_subNode); }
-        iterator & operator ++ () {
-            if (_hashTable->_nodes[_subNode].hasNext()) {
-                _subNode = _hashTable->_nodes[_subNode].getNext();
-            } else {
-                _hash++;
+        iterator(hashtable * hash) : _current(0), _hashTable(hash) {
+            if ((_current < _hashTable->initializedSize()) && ! _hashTable->_nodes[_current].valid()) {
```

Comment:
I think this should be 'current >= size || !valid[current]' to make sure an iterator created from a completely empty hash table matches the 'end' iterator.

Reply:
I did find tests that tested that implicit. I added explicit tests for the same in #4575.
I think it is fine as is.